lib/apns/worker.ex

```diff
@@ -2,336 +2,64 @@ defmodule APNS.Worker do
-        |> state.config.callback_module.error()
-        case rest do
-          "" -> {:noreply, state}
-          _ -> handle_info({:ssl, socket, rest}, %{state | buffer_apple: ""})
-        end
-      buffer ->
-        {:noreply, %{state | buffer_apple: buffer}}
-    end
+    {:noreply, APNS.MessageHandler.handle_response(state, socket, data)}
   end
 
   def handle_info({:ssl, socket, data}, %{socket_feedback: socket} = state) do
-    case <<state.buffer_feedback :: binary, data :: binary>> do
-      <<time :: 32, length :: 16, token :: size(length)-binary, rest :: binary>> ->
-        %APNS.Feedback{time: time, token: Base.encode16(token)}
-        |> state.config.callback_module.feedback()
-        state = %{state | buffer_feedback: ""}
-        case rest do
-          "" -> {:noreply, state}
-          _ -> handle_info({:ssl, socket, rest}, state)
-        end
-      buffer ->
-        {:noreply, %{state | buffer_feedback: buffer}}
-    end
+    {:noreply, APNS.FeedbackHandler.handle_response(state, socket, data)}
   end
 
-  def handle_call(%APNS.Message{token: token} = msg, _from, state) when byte_size(token) != 64 do
-    APNS.Error.new(msg.id, 5)
-    |> state.config.callback_module.error()
-    {:reply, :ok, state}
+  def handle_cast(message, state) do
+    {:noreply, APNS.MessageHandler.push(message, state)}
   end
-  def handle_call(%APNS.Message{} = msg, _from, %{config: config} = state) do
-    limit = case msg.support_old_ios do
-      nil -> config.payload_limit
-      true -> @payload_max_old
-      false -> @payload_max_new
-    end
-    case build_payload(msg, limit) do
-      {:error, :payload_size_exceeded} ->
-        APNS.Error.new(msg.id, 7)
-        |> state.config.callback_module.error()
-        {:reply, :ok, state}
-      payload ->
-        send_message(state.socket_apple, msg, payload)
-        if (state.counter >= state.config.reconnect_after) do
-          Logger.debug "[APNS] #{state.counter} messages sent, reconnecting"
-          send self, :connect_apple
-        end
-        {:reply, :ok, %{state | counter: state.counter + 1}}
-    end
-  end
-
-  def build_payload(msg, payload_limit) do
-    payload = %{aps: %{}}
-
-    if msg.sound do
-      payload = put_in(payload[:aps][:sound], msg.sound)
-    end
-
-    if msg.category != nil do
-      payload = put_in(payload[:aps][:category], msg.category)
-    end
-
-    if msg.badge != nil do
-      payload = put_in(payload[:aps][:badge], msg.badge)
-    end
-
-    if msg.content_available != nil do
-      payload = put_in(payload[:aps][:'content-available'], msg.content_available)
-    end
 
-    if msg.extra != [] do
-      payload = Map.merge(payload, msg.extra)
-    end
-
-    if is_binary(msg.alert) do
-      payload = put_in(payload[:aps][:alert], msg.alert)
-    else
-      payload = put_in(payload[:aps][:alert], format_loc(msg.alert))
-    end
-
-    encode(payload, payload_limit)
+  defp sleep(seconds) do
+    :timer.sleep(seconds * 1000)
```

Comment:
@zepplock, @chvanikoff what do you say? We just moved this from the original code into this file.